### DISCUSSION TITLE
No performance or resource benefit observed in RabbitMQ 3.13 with MQTT

### DISCUSSION THREAD

**Discussion by rajatgl17** [rabbitmq/rabbitmq-server#11754] [Questions] [answered]
As stated in article - https://www.rabbitmq.com/blog/2023/03/21/native-mqtt
starting with version 3.12, RabbitMQ supports Native MQTT, through which it is expected to get results:

1. Memory usage drops by up to 95% and hundreds of GBs with many connections.
2. For the first time ever, RabbitMQ is able to handle millions of connections.

But we don't see any drop in memory usage. Even in version 3.11, memory usage per node was same around 5-6 GB.  Also we are still not able to achieve millions of connections - https://github.com/rabbitmq/rabbitmq-server/discussions/11753

**Data-flow overview**: Mobile application (millions of users) publish click stream events data to our RabbitMQ (version 3.13.3) cluster (behind a AWS network load balancer) with MQTT protocol. We have a cluster of RabbitMQ with 13 nodes (c7.4xlarge). From AMQP topic, data is moved to a random exchange which distributes data to 13 different quorum queues. And from these 13 quorum queues, an application consume data with AMQP 0-9-1 protocol.
All feature flags are enabled on RabbitMQ cluster.

**Comment by ansd** [comment, marked as answer]
The blog post links to https://github.com/ansd/rabbitmq-mqtt/tree/blog-post-native-mqtt so that you can exactly reproduce the results presented in the blog post.

**Comment by michaelklishin** [comment]
> We have a cluster of RabbitMQ with 13 nodes (c7.4xlarge). From AMQP topic, data is moved to a random > exchange which distributes data to 13 different quorum queues

The blog post in question uses MQTT (with an internally used queue type and not AMQP 0-9-1 with quorum queues. The improvements in the MQTT plugin will not reduce memory footprint of quorum queues.
There is [documented tunable parameter](https://www.rabbitmq.com/docs/quorum-queues#resource-use) that can help reduce per quorum queue footprint for mostly idle queues. Millions of quorum queues
would still be a challenge to achieve, resources-wise, regardless of the protocol used.

The blog post never promises that millions of quorum queues will use now have drastically lower memory foorprint.


> But we don't see any drop in memory usage. Even in version 3.11, memory usage per node was same 
> around 5-6 GB. Also we are still not able to achieve millions of connections

We never promised that everyone will benefit from the improvements equally. Neither will be guessing as to what your workload is like. We do not guess in this community.

If you are serious about understanding what is going on in your system, start with [this dedicated section about tuning for a large number of concurrent connections](https://www.rabbitmq.com/docs/networking#tuning-for-large-number-of-connections), then put together an executable way to reproduce your workload.
